parser-typechecker/src/Unison/Codebase/Path.hs

```diff
@@ -71,153 +63,138 @@ import Unison.Prelude hiding (empty, toList)
 
 import Control.Lens hiding (Empty, cons, snoc, unsnoc)
 import qualified Control.Lens as Lens
-import qualified Data.Foldable as Foldable
-import Data.List.Extra (dropPrefix)
 import qualified Data.List.NonEmpty as List.NonEmpty
-import Data.Sequence (Seq ((:<|), (:|>)))
+import Data.Sequence (Seq ((:<|)))
 import qualified Data.Sequence as Seq
 import qualified Data.Text as Text
 import qualified Unison.HashQualified' as HQ'
 import Unison.Name (Convert(..), Name, Parse)
 import qualified Unison.Name as Name
-import Unison.NameSegment (NameSegment (NameSegment))
+import Unison.NameSegment (NameSegment)
 import qualified Unison.NameSegment as NameSegment
 import Unison.Util.Monoid (intercalateMap)
+import Data.Function (on)
+import Unison.Codebase.Position (Position(..))
+import qualified Data.List.Extra as List
+import qualified Data.Foldable as Foldable
 
--- `Foo.Bar.baz` becomes ["Foo", "Bar", "baz"]
-newtype Path = Path { toSeq :: Seq NameSegment } deriving (Eq, Ord, Semigroup, Monoid)
-
-newtype Absolute = Absolute { unabsolute :: Path } deriving (Eq,Ord)
-newtype Relative = Relative { unrelative :: Path } deriving (Eq,Ord)
-newtype Path' = Path' { unPath' :: Either Absolute Relative }
-  deriving (Eq,Ord)
-
-isCurrentPath :: Path' -> Bool
-isCurrentPath p = p == currentPath
-
-currentPath :: Path'
-currentPath = Path' (Right (Relative (Path mempty)))
-
-isRoot' :: Path' -> Bool
-isRoot' = either isRoot (const False) . unPath'
-
-isRoot :: Absolute -> Bool
-isRoot = Seq.null . toSeq . unabsolute
-
-absoluteToPath' :: Absolute -> Path'
-absoluteToPath' abs = Path' (Left abs)
-
-
-
-instance Show Path' where
-  show (Path' (Left abs)) = show abs
-  show (Path' (Right rel)) = show rel
-
-instance Show Absolute where
-  show s = "." ++ show (unabsolute s)
-
-instance Show Relative where
-  show = show . unrelative
-
-unsplit' :: Split' -> Path'
-unsplit' (Path' (Left (Absolute p)), seg) = Path' (Left (Absolute (unsplit (p, seg))))
-unsplit' (Path' (Right (Relative p)), seg) = Path' (Right (Relative (unsplit (p, seg))))
-
-unsplit :: Split -> Path
-unsplit (Path p, a) = Path (p :|> a)
-
-unsplitHQ :: HQSplit -> HQ'.HashQualified Path
-unsplitHQ (p, a) = fmap (snoc p) a
+data Path (pos :: Position) where
+  AbsoluteP :: Seq NameSegment -> Path 'Absolute
+  RelativeP :: Seq NameSegment -> Path 'Relative
+  UncheckedP :: Either (Path 'Absolute) (Path 'Relative) -> Path 'Unchecked
```

Comment:
This constructor just embeds one of the other two constructors. It's a bit weird of course, but is effectively how `Path'` was implemented before, and is hidden from outside callers, who should use the `AbsolutePath` `RelativePath` patterns instead.